Mathlib/Algebra/Ring/Defs.lean

```diff
@@ -0,0 +1,454 @@
+
+This is useful for dealing with submonoids of a ring that contain `-1` without having to duplicate
+lemmas. -/
+class HasDistribNeg (α : Type _) [Mul α] extends HasInvolutiveNeg α where
+  /-- Negation is left distributive over multiplication -/
+  neg_mul : ∀ x y : α, -x * y = -(x * y)
+  /-- Negation is right distributive over multiplication -/
+  mul_neg : ∀ x y : α, x * -y = -(x * y)
+#align has_distrib_neg HasDistribNeg
+
+section Mul
+
+variable [Mul α] [HasDistribNeg α]
+
+@[simp]
+theorem neg_mul (a b : α) : -a * b = -(a * b) :=
+  HasDistribNeg.neg_mul _ _
+#align neg_mul neg_mul
+
+@[simp]
+theorem mul_neg (a b : α) : a * -b = -(a * b) :=
+  HasDistribNeg.mul_neg _ _
+#align mul_neg mul_neg
+
+theorem neg_mul_neg (a b : α) : -a * -b = a * b := by simp
+#align neg_mul_neg neg_mul_neg
+
+theorem neg_mul_eq_neg_mul (a b : α) : -(a * b) = -a * b :=
+  (neg_mul _ _).symm
+#align neg_mul_eq_neg_mul neg_mul_eq_neg_mul
+
+theorem neg_mul_eq_mul_neg (a b : α) : -(a * b) = a * -b :=
+  (mul_neg _ _).symm
+#align neg_mul_eq_mul_neg neg_mul_eq_mul_neg
+
+theorem neg_mul_comm (a b : α) : -a * b = a * -b := by simp
+#align neg_mul_comm neg_mul_comm
+
+end Mul
+
+section MulOneClass
+
+variable [MulOneClass α] [HasDistribNeg α]
+
+theorem neg_eq_neg_one_mul (a : α) : -a = -1 * a := by simp
+
+/-- An element of a ring multiplied by the additive inverse of one is the element's additive
+  inverse. -/
+theorem mul_neg_one (a : α) : a * -1 = -a := by simp
+
+/-- The additive inverse of one multiplied by an element of a ring is the element's additive
+  inverse. -/
+theorem neg_one_mul (a : α) : -1 * a = -a := by simp
+
+end MulOneClass
+
+section MulZeroClass
+
+variable [MulZeroClass α] [HasDistribNeg α]
+
+instance (priority := 100) MulZeroClass.negZeroClass : NegZeroClass α where
+  __ := inferInstanceAs (Zero α); __ := inferInstanceAs (HasInvolutiveNeg α)
+  neg_zero := by rw [← zero_mul (0 : α), ← neg_mul, mul_zero, mul_zero]
+#align mul_zero_class.neg_zero_class MulZeroClass.negZeroClass
+
+end MulZeroClass
+
+end HasDistribNeg
+
+/-!
+### Rings
+-/
+
+
+/-- A not-necessarily-unital, not-necessarily-associative ring. -/
+class NonUnitalNonAssocRing (α : Type u) extends AddCommGroup α, NonUnitalNonAssocSemiring α
+#align non_unital_non_assoc_ring NonUnitalNonAssocRing
+
+-- We defer the instance `non_unital_non_assoc_ring.to_has_distrib_neg` to `algebra.ring.basic`
+-- as it relies on the lemma `eq_neg_of_add_eq_zero_left`.
+/-- An associative but not-necessarily unital ring. -/
+class NonUnitalRing (α : Type _) extends NonUnitalNonAssocRing α, NonUnitalSemiring α
+#align non_unital_ring NonUnitalRing
+
+/-- A unital but not-necessarily-associative ring. -/
+class NonAssocRing (α : Type _) extends NonUnitalNonAssocRing α, NonAssocSemiring α,
+    AddGroupWithOne α
+#align non_assoc_ring NonAssocRing
+
+class Ring (R : Type u) extends Semiring R, AddCommGroup R, AddGroupWithOne R
```

Comment:
This doesn't match the version in Lean 3...

[Zulip Thread](
https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/ring.20vs.20Ring/near/322876462)